public/app/features/query/state/DashboardQueryRunner/DashboardQueryRunner.ts

```diff
@@ -53,44 +53,60 @@ class DashboardQueryRunnerImpl implements DashboardQueryRunner {
-        },
-        error: (err) => subscriber.error(err),
-        complete: () => subscriber.complete(),
-      });
-      return () => {
-        subscription.unsubscribe();
-      };
-    });
+    return this.results.asObservable().pipe(
+      map((result) => {
+        const annotations = getAnnotationsByPanelId(result.annotations, panelId);
+        const alertState = result.alertStates.find((res) => Boolean(panelId) && res.panelId === panelId);
+        return { annotations: dedupAnnotations(annotations), alertState };
+      })
+    );
   }
 
   private executeRun(options: DashboardQueryRunnerOptions) {
     const workers = this.workers.filter((w) => w.canWork(options));
     const workerObservables = workers.map((w) => w.work(options));
-    const observables = [emptyResult()].concat(workerObservables);
-
-    merge(observables)
-      .pipe(
-        takeUntil(this.runs.asObservable()),
-        mergeAll(),
-        reduce((acc, value) => {
-          // should we use scan or reduce here
-          // reduce will only emit when all observables are completed
-          // scan will emit when any observable is completed
-          // choosing reduce to minimize re-renders
-          acc.annotations = acc.annotations.concat(value.annotations);
-          acc.alertStates = acc.alertStates.concat(value.alertStates);
-          return acc;
-        })
-      )
-      .subscribe((x) => {
-        this.results.next(x);
-      });
+
+    const resultSubscription = new Subscription();
+    const resultObservable = merge(workerObservables).pipe(
+      takeUntil(this.runs.asObservable()),
+      mergeAll(),
+      reduce((acc: DashboardQueryRunnerWorkerResult, value: DashboardQueryRunnerWorkerResult) => {
+        // console.log({ acc: acc.annotations.length, value: value.annotations.length });
+        // should we use scan or reduce here
+        // reduce will only emit when all observables are completed
+        // scan will emit when any observable is completed
+        // choosing reduce to minimize re-renders
+        acc.annotations = acc.annotations.concat(value.annotations);
+        acc.alertStates = acc.alertStates.concat(value.alertStates);
+        return acc;
+      }),
+      finalize(() => {
+        resultSubscription.unsubscribe(); // important to avoid memory leaks
+      }),
+      share() // shared because we're using it in takeUntil below
+    );
+
+    const timerSubscription = new Subscription();
+    const timerObservable = timer(200).pipe(
+      mapTo({ annotations: [], alertStates: [] }),
+      takeUntil(resultObservable),
+      finalize(() => {
+        timerSubscription.unsubscribe(); // important to avoid memory leaks
+      })
+    );
+
+    // if the result takes longer than 200ms we just publish an empty result
+    timerSubscription.add(
+      timerObservable.subscribe((result) => {
+        this.results.next(result);
+      })
+    );
```

Comment:
What's the purpose of this empty result after 200ms? we already have a similar thing in runRequest (for the main data queries), is it needed here as well?

Reply:
Is it so that we mergePanelAndDashData & combineLatest get's a result from the DashboardQueryRunner so that we do not block and wait indefinitely on the annotation queries?

Reply:
Exactly, if the dashboard query runner results take more than 200ms we still want to render the panels

Reply:
:+1: